File: cpg-core/src/main/java/de/fraunhofer/aisec/cpg/graph/types/ObjectType.java
/*
 * Copyright (c) 2019, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.graph.types;

import de.fraunhofer.aisec.cpg.frontends.Language;
import de.fraunhofer.aisec.cpg.frontends.LanguageFrontend;
import de.fraunhofer.aisec.cpg.graph.HasType;
import de.fraunhofer.aisec.cpg.graph.Name;
import de.fraunhofer.aisec.cpg.graph.declarations.RecordDeclaration;
import de.fraunhofer.aisec.cpg.graph.edge.Properties;
import de.fraunhofer.aisec.cpg.graph.edge.PropertyEdge;
import java.util.*;
import org.neo4j.ogm.annotation.Relationship;

/**
 * This is the main type in the Type system. ObjectTypes describe objects, as instances of a class.
 * This also includes primitive data types.
 */
public class ObjectType extends Type implements HasType.SecondaryTypeEdge {

  @Override
  public void updateType(Collection<Type> typeState) {
    if (this.generics == null) {
      return;
    }
    for (Type t : this.getGenerics()) {
      for (Type t2 : typeState) {
        if (t2.equals(t)) {
          this.replaceGenerics(t, t2);
        }
      }
    }
  }

  public void replaceGenerics(Type oldType, Type newType) {
    if (this.generics == null) {
      return;
    }
    for (int i = 0; i < this.generics.size(); i++) {
      PropertyEdge<Type> propertyEdge = this.generics.get(i);
      if (propertyEdge.getEnd().equals(oldType)) {
        propertyEdge.setEnd(newType);
      }
    }
  }

  /**
   * ObjectTypes can have a modifier if they are primitive datatypes. The default is signed for
   * primitive data types if there is no more information provided. In case of non-primitive
   * datatypes the modifier is NOT_APPLICABLE
   */
  public enum Modifier {
    SIGNED,
    UNSIGNED,
    NOT_APPLICABLE,
  }

  protected Modifier modifier;
  // Reference from the ObjectType to its class (RecordDeclaration) only if the class is available
  private RecordDeclaration recordDeclaration = null;

  @Relationship(value = "GENERICS", direction = "OUTGOING")
  private List<PropertyEdge<Type>> generics;

  public ObjectType(
      String typeName,
      Qualifier qualifier,
      List<Type> generics,
      Modifier modifier,
      boolean primitive,
      Language<? extends LanguageFrontend> language) {
    super(typeName, qualifier, language);
    this.generics = PropertyEdge.transformIntoOutgoingPropertyEdgeList(generics, this);
    this.modifier = modifier;
    this.primitive = primitive;
    this.setLanguage(language);
  }

  public ObjectType(
      Name typeName,
      Qualifier qualifier,
      List<Type> generics,
      Modifier modifier,
      boolean primitive,
      Language<? extends LanguageFrontend> language) {
    super(typeName, qualifier, language);
    this.generics = PropertyEdge.transformIntoOutgoingPropertyEdgeList(generics, this);
    this.modifier = modifier;
    this.primitive = primitive;
    this.setLanguage(language);
  }

  public ObjectType(
      Type type,
      List<Type> generics,
      Modifier modifier,
      boolean primitive,
      Language<? extends LanguageFrontend> language) {
    super(type);
    this.setLanguage(language);
    this.generics = PropertyEdge.transformIntoOutgoingPropertyEdgeList(generics, this);
    this.modifier = modifier;
    this.primitive = primitive;
  }

  /** Empty default constructor for use in Neo4J persistence. */
  public ObjectType() {
    super();
    this.generics = new ArrayList<>();
    this.modifier = Modifier.NOT_APPLICABLE;
    this.primitive = false;
  }

  public List<Type> getGenerics() {
    List<Type> genericValues = new ArrayList<>();
    for (PropertyEdge<Type> edge : this.generics) {
      genericValues.add(edge.getEnd());
    }
    return Collections.unmodifiableList(genericValues);
  }

  public List<PropertyEdge<Type>> getGenericPropertyEdges() {
    return this.generics;
  }

  public RecordDeclaration getRecordDeclaration() {
    return recordDeclaration;
  }

  public void setRecordDeclaration(RecordDeclaration recordDeclaration) {
    this.recordDeclaration = recordDeclaration;
  }

  /**
   * @return PointerType to a ObjectType, e.g. int*
   */
  @Override
  public PointerType reference(PointerType.PointerOrigin pointerOrigin) {
    return new PointerType(this, pointerOrigin);
  }

  public PointerType reference() {
    return new PointerType(this, PointerType.PointerOrigin.POINTER);
  }

  /**
   * @return UnknownType, as we cannot infer any type information when dereferencing an ObjectType,
   *     as it is just some memory and its interpretation is unknown
   */
  @Override
  public Type dereference() {
    return UnknownType.getUnknownType(getLanguage());
  }

  @Override
  public Type duplicate() {
    ObjectType newObject =
        new ObjectType(this, this.getGenerics(), this.modifier, this.primitive, this.getLanguage());
    newObject.setLanguage(this.getLanguage());
    return newObject;
  }

  public void setGenerics(List<Type> generics) {
    this.generics = PropertyEdge.transformIntoOutgoingPropertyEdgeList(generics, this);
  }

  public void addGeneric(Type generic) {
    var propertyEdge = new PropertyEdge<>(this, generic);
    propertyEdge.addProperty(Properties.INDEX, this.generics.size());
    this.generics.add(propertyEdge);
  }

  public void addGenerics(List<Type> generics) {
    for (Type generic : generics) {
      addGeneric(generic);
    }
  }

  public Modifier getModifier() {
    return modifier;
  }

  @Override
  public boolean isSimilar(Type t) {
    return t instanceof ObjectType
        && this.getGenerics().equals(((ObjectType) t).getGenerics())
        && super.isSimilar(t);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (!(o instanceof ObjectType)) return false;
    if (!super.equals(o)) return false;
    ObjectType that = (ObjectType) o;
    return Objects.equals(this.getGenerics(), that.getGenerics())
        && PropertyEdge.propertyEqualsList(generics, that.generics)
        && this.primitive == that.primitive
        && this.modifier.equals(that.modifier);
  }

  @Override
  public int hashCode() {
    return Objects.hash(super.hashCode(), generics, modifier, primitive);
  }
}

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/types/ObjectType.kt
/*
 * Copyright (c) 2019, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.graph.types

import de.fraunhofer.aisec.cpg.PopulatedByPass
import de.fraunhofer.aisec.cpg.frontends.Language
import de.fraunhofer.aisec.cpg.graph.declarations.ConstructorDeclaration
import de.fraunhofer.aisec.cpg.graph.declarations.Declaration
import de.fraunhofer.aisec.cpg.graph.declarations.FieldDeclaration
import de.fraunhofer.aisec.cpg.graph.declarations.MethodDeclaration
import de.fraunhofer.aisec.cpg.graph.declarations.RecordDeclaration
import de.fraunhofer.aisec.cpg.graph.scopes.Scope
import de.fraunhofer.aisec.cpg.graph.types.PointerType.PointerOrigin
import de.fraunhofer.aisec.cpg.graph.unknownType
import de.fraunhofer.aisec.cpg.helpers.identitySetOf
import de.fraunhofer.aisec.cpg.passes.TypeResolver
import java.util.*
import org.neo4j.ogm.annotation.Relationship

/**
 * This is the main type in the Type system. ObjectTypes describe objects, as instances of a class.
 * This also includes primitive data types.
 */
open class ObjectType : Type, HasSecondaryTypeEdge {
    /**
     * Reference from the [ObjectType] to its class ([RecordDeclaration]), only if the class is
     * available. This is set by the [TypeResolver].
     *
     * This also sets this type's [scope] to the [RecordDeclaration.scope].
     */
    @PopulatedByPass(TypeResolver::class)
    var recordDeclaration: RecordDeclaration? = null
        set(value) {
            field = value
            this.scope = value?.scope
        }

    @Relationship(value = "GENERICS", direction = Relationship.Direction.OUTGOING)
    var generics: List<Type>
        private set

    constructor(
        typeName: CharSequence,
        generics: List<Type>,
        primitive: Boolean,
        mutable: Boolean,
        language: Language<*>,
    ) : super(typeName, language) {
        this.generics = generics
        this.isPrimitive = primitive
        this.isMutable = mutable
        this.language = language
    }

    constructor(
        typeName: CharSequence,
        generics: List<Type>,
        primitive: Boolean,
        language: Language<*>,
    ) : super(typeName, language) {
        this.generics = generics
        this.isPrimitive = primitive
        this.isMutable = true
        this.language = language
    }

    /** Empty default constructor for use in Neo4J persistence. */
    constructor() : super() {
        this.isPrimitive = false
        this.isMutable = true
        this.generics = mutableListOf<Type>()
    }

    /** @return PointerType to a ObjectType, e.g. int* */
    override fun reference(pointer: PointerOrigin?): PointerType {
        return PointerType(this, pointer)
    }

    fun reference(): PointerType {
        return PointerType(this, PointerOrigin.POINTER)
    }

    /**
     * @return UnknownType, as we cannot infer any type information when de-referencing an
     *   ObjectType, as it is just some memory and its interpretation is unknown
     */
    override fun dereference(): Type {
        return unknownType()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ObjectType) return false
        if (!super.equals(other)) return false
        return generics == other.generics && isPrimitive == other.isPrimitive
    }

    override fun hashCode() = Objects.hash(super.hashCode(), generics, isPrimitive)

    override val secondaryTypes: List<Type>
        get() = generics

    /**
     * Returns all constructors that are declared in this type and its super types. See
     * [findMembers] for more details.
     */
    val constructors: Set<ConstructorDeclaration>
        get() {
            return findMembers<ConstructorDeclaration>()
        }

    /**
     * Returns all methods that are declared in this type and its super types. See [findMembers] for
     * more details.
     */
    val methods: Set<MethodDeclaration>
        get() {
            return findMembers<MethodDeclaration>()
        }

    /**
     * Returns all fields that are declared in this type and its super types. See [findMembers] for
     * more details.
     */
    val fields: Set<FieldDeclaration>
        get() {
            return findMembers<FieldDeclaration>()
        }

    /**
     * Returns all [Declaration] nodes (of type [T]) that are declared in this type and its super
     * types. We use the underlying [recordDeclaration] of the type to find the [Scope] it declares
     * and then look for appropriate symbols pointing to a [Declaration].
     */
    private inline fun <reified T : Declaration> findMembers(): Set<T> {
        // We need to gather all members that are in within the scope of the underlying record
        // declaration, as well as their super types
        val members = mutableSetOf<T>()

        // Gather all members of this type and its super-types
        val worklist = mutableListOf<Type>(this)
        val alreadySeen = identitySetOf<Type>()
        while (worklist.isNotEmpty()) {
            val next = worklist.removeFirst()

            // Add all members in the declaring scope
            next.recordDeclaration
                ?.declaringScope
                ?.symbols
                ?.values
                ?.flatten()
                ?.filterIsInstanceTo(members)

            // Add super types
            worklist += next.superTypes
            alreadySeen.add(next)
        }

        return members.toSet()
    }
}
